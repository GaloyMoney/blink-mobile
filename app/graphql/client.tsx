import { GraphQLWsLink } from "@apollo/client/link/subscriptions"
import { createClient } from "graphql-ws"

import {
  ApolloClient,
  ApolloLink,
  ApolloProvider,
  HttpLink,
  NormalizedCacheObject,
  ServerError,
  split,
} from "@apollo/client"
import AsyncStorage from "@react-native-async-storage/async-storage"
import DeviceInfo from "react-native-device-info"

import { setContext } from "@apollo/client/link/context"
import { RetryLink } from "@apollo/client/link/retry"
import { getMainDefinition } from "@apollo/client/utilities"

import { createPersistedQueryLink } from "@apollo/client/link/persisted-queries"
import { BUILD_VERSION } from "@app/config"
import { useAppConfig } from "@app/hooks"
import { AsyncStorageWrapper, CachePersistor } from "apollo3-cache-persist"
import jsSha256 from "js-sha256"
import React, { PropsWithChildren, useEffect, useState } from "react"
import { createCache } from "./cache"

import { useI18nContext } from "@app/i18n/i18n-react"
import { isIos } from "../utils/helper"
import { loadString, saveString } from "../utils/storage"
import { AnalyticsContainer } from "./analytics"
import { useLanguageQuery, useRealtimePriceQuery } from "./generated"
import { IsAuthedContextProvider, useIsAuthed } from "./is-authed-context"

import { onError } from "@apollo/client/link/error"

import { getLanguageFromString, getLocaleFromLanguage } from "@app/utils/locale-detector"
import { NetworkErrorToast } from "./network-error-toast"
import { MessagingContainer } from "./messaging"

const noRetryOperations = [
  "intraLedgerPaymentSend",
  "intraLedgerUsdPaymentSend",

  "lnInvoiceFeeProbe",
  "lnInvoicePaymentSend",
  "lnNoAmountInvoiceFeeProbe",
  "lnNoAmountInvoicePaymentSend",
  "lnNoAmountUsdInvoiceFeeProbe",
  "lnUsdInvoiceFeeProbe",
  "lnNoAmountUsdInvoicePaymentSend",

  "onChainPaymentSend",
  "onChainTxFee",

  // no need to retry to upload the token
  // specially as it's running on app start
  // and can create some unwanted loop when token is not valid
  "deviceNotificationTokenCreate",
]

const getAuthorizationHeader = (token: string): string => {
  return `Bearer ${token}`
}

const GaloyClient: React.FC<PropsWithChildren> = ({ children }) => {
  const { appConfig, saveToken } = useAppConfig()
  const [networkError, setNetworkError] = useState<ServerError>()

  const [apolloClient, setApolloClient] = useState<{
    client: ApolloClient<NormalizedCacheObject>
    isAuthed: boolean
  }>()

  useEffect(() => {
    ;(async () => {
      const token = appConfig.token

      console.log(
        `creating new apollo client, token: ${Boolean(token)}, uri: ${
          appConfig.galoyInstance.graphqlUri
        }`,
      )

      const httpLink = new HttpLink({
        uri: appConfig.galoyInstance.graphqlUri,
      })

      // persistedQuery provide client side bandwidth optimization by returning a hash
      // of the uery instead of the whole query
      //
      // use the following line if you want to deactivate in dev
      // const persistedQueryLink = httpLink
      //
      // we are using "js-sha256" because crypto-hash has compatibility issue with react-native
      // from "@apollo/client/link/persisted-queries/types" but not exporterd
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      type SHA256Function = (...args: any[]) => string | PromiseLike<string>
      const sha256: SHA256Function = jsSha256 as unknown as SHA256Function
      const persistedQueryLink = createPersistedQueryLink({ sha256 })

      const authLink = setContext((request, { headers }) => ({
        headers: {
          ...headers,
          authorization: token ? getAuthorizationHeader(token) : "",
        },
      }))

      const retryLink = new RetryLink({
        attempts: {
          max: 5,
          retryIf: (error, operation) => {
            console.debug(JSON.stringify(error), "retry on error")
            return Boolean(error) && !noRetryOperations.includes(operation.operationName)
          },
        },
      })

      const wsLink = new GraphQLWsLink(
        createClient({
          url: appConfig.galoyInstance.graphqlWsUri,
          connectionParams: token
            ? { Authorization: getAuthorizationHeader(token) }
            : undefined,
          retryAttempts: 12,
          shouldRetry: (errOrCloseEvent) => {
            console.warn(
              { errOrCloseEvent },
              "entering shouldRetry function for websocket",
            )
            // TODO: understand how the backend is closing the connection
            // for instance during a new version rollout or k8s upgrade
            //
            // in the meantime:
            // returning true instead of the default 'Any non-`CloseEvent`'
            // to force createClient to attempt a reconnection
            return true
          },
          // Voluntary not using: webSocketImpl: WebSocket
          // seems react native already have an implement of the websocket?
          //
          // TODO: implement keepAlive and reconnection?
          // https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#keepalive
        }),
      )

      const errorLink = onError(({ graphQLErrors, networkError }) => {
        // graphqlErrors should be managed locally
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) => {
            if (message === "PersistedQueryNotFound") {
              console.log(`[GraphQL info]: Message: ${message}, Path: ${path}}`, {
                locations,
              })
            } else {
              console.warn(`[GraphQL error]: Message: ${message}, Path: ${path}}`, {
                locations,
              })
            }
          })
        // only network error are managed globally
        if (networkError) {
          console.log(`[Network error]: ${networkError}`)
          setNetworkError(networkError as ServerError)
        }
      })

      const link = split(
        ({ query }) => {
          const definition = getMainDefinition(query)
          return (
            definition.kind === "OperationDefinition" &&
            definition.operation === "subscription"
          )
        },
        wsLink,
        ApolloLink.from([retryLink, authLink, persistedQueryLink, errorLink, httpLink]),
      )

      const cache = createCache()

      const persistor = new CachePersistor({
        cache,
        storage: new AsyncStorageWrapper(AsyncStorage),
        debug: __DEV__,

        persistenceMapper: async (_data) => {
          // TODO:
          // we should only store the last 20 transactions to keep the cache small
          // there could be other data to filter as well
          // filter your cached data and queries
          // return filteredData
          return _data
        },
      })

      const readableVersion = DeviceInfo.getReadableVersion()

      const client = new ApolloClient({
        cache,
        link,
        name: isIos ? "iOS" : "Android",
        version: readableVersion,
        connectToDevTools: true,
      })

      const buildVersion = DeviceInfo.getBuildNumber()

      // Read the current version from AsyncStorage.
      const currentVersion = await loadString(BUILD_VERSION)

      if (currentVersion === buildVersion) {
        // If the current version matches the latest version,
        // we're good to go and can restore the cache.
        await persistor.restore()
      } else {
        // Otherwise, we'll want to purge the outdated persisted cache
        // and mark ourselves as having updated to the latest version.

        // init the DB. will be override if a cache exists
        await persistor.purge()
        await saveString(BUILD_VERSION, buildVersion)
      }

      client.onClearStore(persistor.purge)

      setApolloClient({ client, isAuthed: Boolean(token) })

      return () => client.cache.reset()
    })()
  }, [appConfig.token, appConfig.galoyInstance, saveToken])

  // Before we show the app, we have to wait for our state to be ready.
  // In the meantime, don't render anything. This will be the background
  // color set in native by rootView's background color.
  //
  // This step should be completely covered over by the splash screen though.
  //
  // You're welcome to swap in your own component to render if your boot up
  // sequence is too slow though.
  if (!apolloClient) {
    return <></>
  }

  return (
    <ApolloProvider client={apolloClient.client}>
      <IsAuthedContextProvider value={apolloClient.isAuthed}>
        <MessagingContainer />
        <NetworkErrorToast networkError={networkError} />
        <LanguageSync />
        <AnalyticsContainer />
        <MyPriceUpdates />
        {children}
      </IsAuthedContextProvider>
    </ApolloProvider>
  )
}

const MyPriceUpdates = () => {
  const isAuthed = useIsAuthed()

  const pollInterval = 5 * 60 * 1000 // 5 min
  useRealtimePriceQuery({
    // only fetch after pollInterval
    // the first query is done by the home page automatically
    fetchPolicy: "cache-only",
    nextFetchPolicy: "network-only",
    pollInterval,
    skip: !isAuthed,
  })

  return null
}

const LanguageSync = () => {
  const isAuthed = useIsAuthed()

  const { data } = useLanguageQuery({ fetchPolicy: "cache-first", skip: !isAuthed })

  const userPreferredLocale = getLocaleFromLanguage(
    getLanguageFromString(data?.me?.language),
  )
  const { locale, setLocale } = useI18nContext()

  useEffect(() => {
    if (userPreferredLocale !== locale) {
      setLocale(userPreferredLocale)
    }
    // setLocale is not set as a dependency because it changes every render
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [userPreferredLocale, locale])

  return <></>
}

export { GaloyClient }
