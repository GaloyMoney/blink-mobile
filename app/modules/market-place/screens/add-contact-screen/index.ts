export * from './add-contact-screen'