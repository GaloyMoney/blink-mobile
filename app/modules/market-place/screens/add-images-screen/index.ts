export * from './add-images-screen'