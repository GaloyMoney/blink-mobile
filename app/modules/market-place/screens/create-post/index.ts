export * from './create-post-screen';