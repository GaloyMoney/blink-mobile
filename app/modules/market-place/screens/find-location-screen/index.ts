export * from './find-location-screen'