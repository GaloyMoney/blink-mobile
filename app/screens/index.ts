export * from "./backup-screen"
export * from "./import-wallet-screen"
export * from "./transaction-history"
export * from "./intro-screen"
