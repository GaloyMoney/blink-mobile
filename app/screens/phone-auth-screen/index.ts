export * from "./phone-login-validation"
export * from "./phone-login-input"
