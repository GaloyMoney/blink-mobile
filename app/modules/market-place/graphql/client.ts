import {
  ApolloClient,
  ApolloLink,
  DefaultOptions,
  from,
  HttpLink,
  InMemoryCache,
  fromPromise,
} from "@apollo/client"

import { getMainDefinition } from "@apollo/client/utilities"
import { setContext } from "@apollo/client/link/context"
import { onError } from "@apollo/client/link/error"
import { createUploadLink } from "apollo-upload-client"
import { scriptHostname } from "../utils/helper"

const attemps = 0
const GRAPHQL_REGTEST_URI = `http://${scriptHostname()}:4002/graphql`
const GRAPHQL_TESTNET_URI = "https://api.pvbtc1.staging.pvbtc.cloud/graphql"
const GRAPHQL_MAINNET_URI = "https://api.production.pvbtc.cloud/graphql"

const GRAPHQL_REGTEST_WS_URI = `ws://${scriptHostname()}:4002/graphql`
const GRAPHQL_TESTNET_WS_URI = "wss://api.pvbtc1.staging.pvbtc.cloud/graphql"
const GRAPHQL_MAINNET_WS_URI = "wss://api.production.pvbtc.cloud/graphql"
const errorLink = onError(({ graphQLErrors, networkError, operation }) => {
  if (graphQLErrors)
    graphQLErrors.forEach(async ({ message, locations, path }) => {
      switch (message) {
        case "Unauthorized": {
        }
        default:
          console.log(``)
      }
    })
  if (networkError) console.log(`[Network error]: ${networkError}`)
})
const httpLink = createUploadLink({ uri: "https://api-puravida.herokuapp.com/graphql" })
// new HttpLink({
//   uri: `https://api-puravida.herokuapp.com/graphql`,
// })

const authLink = setContext(async (_, { headers }) => {
  // get the authentication token from local storage if it exists
  // const token = localStorage.getItem('token');
  const token = ""
  // return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      token: token ? `${token}` : "",
    },
  }
})

export const cache = new InMemoryCache({ addTypename: false })

const defaultOptions: DefaultOptions = {
  watchQuery: {
    fetchPolicy: "no-cache",
    errorPolicy: "ignore",
  },
  query: {
    fetchPolicy: "no-cache",
    errorPolicy: "all",
  },
}

const client = new ApolloClient({
  // link: authLink.concat(errorLink.concat(Â©httpLink)),
  link: from([authLink, errorLink, httpLink]),
  cache,
  defaultOptions,
})

export default client
