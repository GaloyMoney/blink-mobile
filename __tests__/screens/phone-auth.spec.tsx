import * as React from "react"
import { MockedProvider } from "@apollo/client/testing"
import { InMemoryCache } from "@apollo/client"
import { act, cleanup, fireEvent, render } from "@testing-library/react-native"
import "@testing-library/jest-native/extend-expect"
import "react-native-gesture-handler/jestSetup.js"
import { translateUnknown as translate } from "@galoymoney/client"
import "@mocks/react-navigation-native"
import "@mocks/react-native-geetest-module"
import { WelcomePhoneInputScreen } from "@app/screens/phone-auth-screen"
import {
  AppConfiguration,
  AppConfigurationContext,
  defaultConfiguration,
} from "../../app/context/app-configuration"

jest.mock("react-native/Libraries/EventEmitter/NativeEventEmitter")
jest.mock("react-native-fingerprint-scanner", () => ({}))

const cache = new InMemoryCache()

describe("WelcomePhoneInputScreen", () => {
  afterEach(cleanup)
  it("render matches snapshot", () => {
    const tree = render(
      <AppConfigurationContext.Provider
        value={{
          appConfig: defaultConfiguration,
          setAppConfig: (config: AppConfiguration) => {},
        }}
      >
        <MockedProvider cache={cache}>
          <WelcomePhoneInputScreen />
        </MockedProvider>
      </AppConfigurationContext.Provider>,
    )
    expect(tree).toMatchSnapshot()
  })
  it("has TextInput", () => {
    const { queryByA11yLabel, queryByPlaceholderText } = render(
      <AppConfigurationContext.Provider
        value={{
          appConfig: defaultConfiguration,
          setAppConfig: (config: AppConfiguration) => {},
        }}
      >
        <MockedProvider cache={cache}>
          <WelcomePhoneInputScreen />
        </MockedProvider>
      </AppConfigurationContext.Provider>,
    )
    expect(queryByA11yLabel("Input phone number")).not.toBeNull()
    expect(
      queryByPlaceholderText(translate("WelcomePhoneInputScreen.placeholder")),
    ).not.toBeNull()
  })
  it("country picker is visible on press", async () => {
    const { getByTestId } = render(
      <AppConfigurationContext.Provider
        value={{
          appConfig: defaultConfiguration,
          setAppConfig: (config: AppConfiguration) => {},
        }}
      >
        <MockedProvider cache={cache}>
          <WelcomePhoneInputScreen />
        </MockedProvider>
      </AppConfigurationContext.Provider>,
    )

    const DropDownButton = getByTestId("DropDownButton")
    await act(() => {
      fireEvent.press(DropDownButton)
    })
    const countryPicker = getByTestId("country-picker")
    expect(countryPicker).toHaveProp("visible", true)
  })
})
