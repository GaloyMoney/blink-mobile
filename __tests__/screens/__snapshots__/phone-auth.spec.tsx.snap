// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WelcomePhoneInputScreen render matches snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Array [
        Object {
          "backgroundColor": "#E6EBEf",
          "flex": 1,
          "height": "100%",
        },
        Object {
          "backgroundColor": "#E6EBEf",
        },
      ],
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection={
      Array [
        "down",
      ]
    }
    swipeThreshold={50}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      forwardedRef={[Function]}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      collapsable={false}
      deviceHeight={null}
      deviceWidth={null}
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      nativeID="animatedComponent"
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onSwipeComplete={[Function]}
      panResponderThreshold={4}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      statusBarTranslucent={false}
      style={
        Object {
          "flex": 1,
          "flexGrow": 1,
          "justifyContent": "center",
          "left": 0,
          "margin": 37.5,
          "marginBottom": 0,
          "marginHorizontal": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection={
        Array [
          "down",
        ]
      }
      swipeThreshold={50}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "space-around",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </View>
      <RNCSafeAreaView
        edges={
          Array [
            "bottom",
            "left",
            "right",
            "top",
          ]
        }
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "paddingHorizontal": 20,
            "paddingTop": 10,
          }
        }
      >
        <View
          style={
            Object {
              "height": 14,
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#CFD9E2",
                  "fontSize": 72,
                },
                Object {
                  "height": 40,
                  "top": -40,
                },
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            ï“‰
          </Text>
        </View>
        <Text
          style={
            Object {
              "color": "#1d1d1d",
              "fontSize": 18,
              "marginVertical": 8,
            }
          }
        >
          
        </Text>
        <View
          style={
            Object {
              "alignContent": "stretch",
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderRadius": 3,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 3,
                },
                undefined,
                false,
              ]
            }
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "busy": false,
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#3050C4",
                    "borderColor": "#2089dc",
                    "borderRadius": 3,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginHorizontal": 20,
                    "marginVertical": 10,
                    "padding": 8,
                    "width": 145,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 18,
                      "paddingVertical": 1,
                      "textAlign": "center",
                    }
                  }
                >
                  Dismiss
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderRadius": 3,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 3,
                },
                undefined,
                false,
              ]
            }
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "busy": false,
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#3050C4",
                    "borderColor": "#2089dc",
                    "borderRadius": 3,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginHorizontal": 20,
                    "marginVertical": 10,
                    "padding": 8,
                    "width": 145,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 18,
                      "paddingVertical": 1,
                      "textAlign": "center",
                    }
                  }
                >
                  Open
                </Text>
              </View>
            </View>
          </View>
        </View>
      </RNCSafeAreaView>
    </View>
  </Modal>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#E6EBEf",
          "flex": 1,
          "height": "100%",
        },
        Object {
          "backgroundColor": "#E6EBEf",
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Array [
          Object {
            "alignItems": "stretch",
            "justifyContent": "flex-start",
          },
          Object {},
        ]
      }
      style={
        Array [
          Object {
            "backgroundColor": "#E6EBEf",
            "flex": 1,
            "height": "100%",
          },
          Object {
            "backgroundColor": "#E6EBEf",
          },
        ]
      }
    >
      <View>
        <View
          style={
            Object {
              "flex": 1,
              "justifyContent": "space-around",
              "marginTop": 50,
            }
          }
        >
          <View>
            < />
            <Text>
              Enter your phone number, and we'll text you an access code.
            </Text>
          </View>
          <View
            onLayout={[Function]}
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "flexDirection": "row",
                    "width": 600,
                  },
                  Object {},
                  Object {},
                ]
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                nativeID="animatedComponent"
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "height": "100%",
                    "justifyContent": "center",
                    "minWidth": 32,
                    "opacity": 1,
                    "width": 150,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                      "marginRight": 10,
                      "width": 30,
                    }
                  }
                >
                  <ActivityIndicator
                    size="small"
                  />
                </View>
                <Modal
                  animated={true}
                  animationType="slide"
                  hardwareAccelerated={false}
                  onDismiss={[Function]}
                  onRequestClose={[Function]}
                  testID="country-picker"
                  visible={false}
                >
                  <RCTSafeAreaView
                    emulateUnlessSupported={true}
                    style={
                      Array [
                        Object {
                          "flex": 1,
                        },
                        Object {
                          "backgroundColor": "#ffffff",
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "height": 48,
                              "justifyContent": "center",
                              "width": "15%",
                            },
                            undefined,
                          ]
                        }
                      >
                        <View
                          accessible={true}
                          collapsable={false}
                          focusable={true}
                          nativeID="animatedComponent"
                          onClick={[Function]}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            Object {
                              "opacity": 1,
                            }
                          }
                        >
                          <Image
                            source={
                              Object {
                                "testUri": "../../react-native-country-picker-modal/lib/assets/images/close.ios.png",
                              }
                            }
                            style={
                              Array [
                                Object {
                                  "height": 25,
                                  "resizeMode": "contain",
                                  "width": 25,
                                },
                                undefined,
                                Object {
                                  "tintColor": "#000000",
                                },
                              ]
                            }
                          />
                        </View>
                      </View>
                      <TextInput
                        autoCorrect={false}
                        autoFocus={false}
                        onBlur={[Function]}
                        onChangeText={[Function]}
                        onClose={[Function]}
                        onFocus={[Function]}
                        placeholder="Enter country name"
                        placeholderTextColor="#aaa"
                        renderFilter={[Function]}
                        style={
                          Array [
                            Object {
                              "height": 48,
                              "width": "70%",
                            },
                            Object {
                              "color": "#000000",
                              "fontFamily": "System",
                              "fontSize": 16,
                            },
                          ]
                        }
                        testID="text-input-country-filter"
                        value=""
                        withCloseButton={true}
                        withFilter={true}
                      />
                    </View>
                    <View
                      style={
                        Array [
                          Object {
                            "flex": 1,
                            "flexDirection": "row",
                            "justifyContent": "space-between",
                          },
                          Object {
                            "backgroundColor": "#ffffff",
                          },
                        ]
                      }
                    >
                      <RCTScrollView
                        ItemSeparatorComponent={[Function]}
                        automaticallyAdjustContentInsets={false}
                        data={Array []}
                        getItem={[Function]}
                        getItemCount={[Function]}
                        getItemLayout={[Function]}
                        initialNumToRender={14}
                        keyExtractor={[Function]}
                        keyboardShouldPersistTaps="handled"
                        onContentSizeChange={[Function]}
                        onLayout={[Function]}
                        onMomentumScrollBegin={[Function]}
                        onMomentumScrollEnd={[Function]}
                        onScroll={[Function]}
                        onScrollBeginDrag={[Function]}
                        onScrollEndDrag={[Function]}
                        onScrollToIndexFailed={[Function]}
                        removeClippedSubviews={false}
                        renderItem={[Function]}
                        scrollEventThrottle={1}
                        stickyHeaderIndices={Array []}
                        testID="list-countries"
                        viewabilityConfigCallbackPairs={Array []}
                      >
                        <View />
                      </RCTScrollView>
                    </View>
                  </RCTSafeAreaView>
                </Modal>
                <
                  height={14}
                  testID="DropDownButton"
                  width={12}
                />
              </View>
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#F8F9F9",
                      "flex": 1,
                      "flexDirection": "row",
                      "paddingHorizontal": 30,
                      "paddingVertical": 27,
                      "textAlign": "left",
                    },
                    Object {},
                  ]
                }
              >
                <TextInput
                  accessibilityLabel="Input phone number"
                  autoFocus={true}
                  editable={true}
                  keyboardAppearance="default"
                  keyboardType="phone-pad"
                  onChangeText={[Function]}
                  onSubmitEditing={[Function]}
                  placeholder="Phone Number"
                  returnKeyType="done"
                  selectionColor="black"
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "flex": 1,
                        "fontSize": 16,
                      },
                      Object {},
                    ]
                  }
                  textContentType="telephoneNumber"
                  value=""
                />
              </View>
            </View>
            <ActivityIndicator
              animating={false}
              color="#3050C4"
              size="large"
              style={
                Object {
                  "marginTop": 32,
                }
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "borderRadius": 3,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 3,
                },
                undefined,
                false,
              ]
            }
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "busy": false,
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              nativeID="animatedComponent"
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#2089dc",
                    "borderColor": "#2089dc",
                    "borderRadius": 3,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 8,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 18,
                      "paddingVertical": 1,
                      "textAlign": "center",
                    }
                  }
                >
                  Continue
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View>
          <Text
            allowFontScaling={false}
            onPress={[Function]}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#1d1d1d",
                  "fontSize": 12,
                },
                undefined,
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            ïˆ£
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
